tokio/src/sync/rwlock/owned_write_guard.rs

```diff
@@ -0,0 +1,107 @@
+use crate::sync::rwlock::owned_read_guard::OwnedRwLockReadGuard;
+use crate::sync::rwlock::RwLock;
+use std::fmt;
+use std::mem::{self, ManuallyDrop};
+use std::ops;
+use std::sync::Arc;
+
+/// Owned RAII structure used to release the exclusive write access of a lock when
+/// dropped.
+///
+/// This structure is created by the [`write_owned`] method
+/// on [`RwLock`].
+///
+/// [`write_owned`]: method@crate::sync::RwLock::write_owned
+/// [`RwLock`]: struct@crate::sync::RwLock
+pub struct OwnedRwLockWriteGuard<T: ?Sized> {
+    // ManuallyDrop allows us to destructure into this field without running the destructor.
+    pub(super) lock: ManuallyDrop<Arc<RwLock<T>>>,
+    pub(super) data: *mut T,
+}
```

Comment:
This struct doesn't have any `map` methods, but the non-owned variant does. Note that mapping a write guard should result in a separate type.

Reply:
I mentioned why I didn't add them in the commit message, but essentially because it stores a pointer to the actual `RwLock` instead of the internal `Semaphore` it has to keep hold of the original generic parameter even after maps. So to support this there would have to be separate mapped variants with two generic parameters instead of one. Alternatively `OwnedRwLockWriteGuard` could have two generic parameters - which do you think is better?

Reply:
We have to have a mapped variant _anyway_, at least for the write variant. I suppose it would be ok to do the same for the read variant.

Reply:
Actually, I just realized it is possible to have a `struct OwnedRwLockReadGuard<T, U = T>`, I'll do that instead.